README

  Section of Biomedical Image Analysis
  Department of Radiology
  University of Pennsylvania
  3600 Market Street, Suite 380
  Philadelphia, PA 19104

  Web:   https://www.rad.upenn.edu/sbia/
  Email: sbia-software -at- uphs.upenn.edu

  Copyright (c) 2011 University of Pennsylvania. All rights reserved.
  See COPYING file or https://www.rad.upenn.edu/sbia/software/license.html.



INTRODUCTION
============

  The Build system And Software Implementation Standard (BASIS) software package
  not only provides a project template which standardizes the project directory
  structure, it also accompanies this template with CMake implementations.
  These CMake implementations ease and standardize the build, testing,
  packaging, and installation of software packages developed at SBIA.
  Besides the software components, the documentation of BASIS defines the
  lab standard for various aspects of a software development project.
  Moreover, fundamental utility source files and libraries written both at
  SBIA and external sites are integrated as components of BASIS.

  Every SBIA software project which was started after the introduction of BASIS
  in June 2011 is required to built on top of BASIS, i.e., to instantiate the
  project template and utilize the provided CMake implementations. An extensive
  documentation is provided with BASIS to ease this task as much as possible.

  The BASIS package, when build and installed once globally on a target system
  on which other SBIA software is to be built, is used for the build of these
  dependent SBIA projects. Alternatively, if the package is not found, each SBIA
  project which is built on top of BASIS and implements the super-build feature,
  retrieves and builds a local copy using CMake's ExternalProject module
  (http://www.kitware.com/products/html/BuildingExternalProjectsWithCMake2.8.html).

  Another feature of BASIS is, that it not only eases the developement and
  maintenance of separatly managed software projects, it also enables the fusion
  of these more or less independently developed software packages into so-called
  superprojects following the super-build methodology which more recently gained
  more and more attraction on the CMake community. In this context, the separately
  managed software packages are considered components of the superproject.



PACKAGE OVERVIEW
================

  CMake Modules
  -------------

  The CMake modules and corresponding auxiliary files are not only used by the
  BASIS package itself, but also any SBIA project which follows BASIS for the
  configuration of the CMake-based build system. These CMake modules in
  particular define many functions which are used as replacement of CMake's
  built-in functions. For example, the function basis_add_executable () is used
  by SBIA projects as replacement for CMake's add_executable () command.
  The SBIA CMake functions are convenient to use, extend the functionality of
  CMake's built-in functions and enable BASIS to impose its standard on the
  configured build systems.


  Project Template
  ----------------

  Every SBIA project that wants to follow BASIS has to be derived from the BASIS
  Project Template. For convenience, a project creation script is provided which
  automates the instantiation of the template for new projects.



ChangeLog.txt
.. This file is formatted using reStructuredText (reST) and included by doc/changelog.rst !

.. _v3.3.1:

Release 3.3.1 (Feb 2, 2017)
---------------------------

- doc: Update Quickstart documentation
- fix: Support package names with dashes
- fix: Build of Python script with future statements
- fix: Ensure that order of `basis_install` arguments does not matter
- fix: Use `BASIS_GE_CONFIG` instead of (deprecated) `$<CONFIGURATION>`
- fix: `BASIS_DIR` used to configure test projects
- fix: Consider `$<CONFIG>` subdirectory in executable target info path
- fix: Use per-configuration Doxygen configuration file with `$<CONFIG>` replaced
- fix: Have Doxygen ignore script target `<target_build_tmp>.dir/build/` directories
- fix: Build of Python script with future statements
- fix: Skip `list(REMOVE_ITEMS)` in `basis_finalize_targets` when `TARGETS` list empty
- fix: Make `[PROJECT_]OTHER_DIRS` absolute
- fix: Duplicate ‘lib’ target issue
- fix: When looking for multiple package versions, look in user `DEPENDS` path first
- del: Remove obsolete `FindOpenCV` module


.. _v3.3.0:

Release 3.3.0 (Apr 19, 2016)
----------------------------

- Changed license to standard BSD 2-clause License without SBIA specific terms.
- Moved project to new GitHub repositories with big objects from history removed.
- Moved SBIA project template to separate repository.
- CMake BASIS Modules and Find modules available in separate Git repositories.
- Enhanced project modularization support.
- Add C++11 compiler flags by specifying ``LANGUAGES CXX-11`` in ``BasisProject.cmake`` file.
- Added Find modules for: TBB, ARPACK, UMFPACK, Eigen3, FLANN, Python, Jython.
- Added ``basis_project`` attributes: ``SOVERSION``, ``DEFAULT_MODULES``, ``TOOLS_DIR``, ``OTHER_DIRS``.
- Support separate ``CODE_DIR`` and ``TOOLS_DIR`` for library and application code.
- Added ``BUILD_APPLICATIONS`` option when ``TOOLS_DIR/CMakeLists.txt`` exists.
- Added ``HAVE_<Package>`` and ``HAVE_<Package>_<Module>`` defines.
- Replace ``USE_<Package>`` by more common ``WITH_<Package>`` options.
- Changed order of dependencies discovery and inclusion of ``Depends.cmake`` file
- Use file of other module included when declared as dependency of a module.
- Modified ``basis_add_subdirectory`` to support specification of module directory.
- Added ``basis_find_package`` option: ``OPTIONAL_COMPONENTS``.
- Modified ``basis_find_package`` to add ``DEPENDS_<Package>_DIR`` cache entry
  instead of ``<Package>_DIR`` for either root installation directory or location of
  ``<Package>Config.cmake`` file. Allow alternative ``cmake`` command-line arguments
  such as ``<Package>_ROOT`` and ``<Package>_DIR`` to initialize package search paths.
- Support specification of alternative package versions of dependencies.
- Declare dependencies on other modules using ``<Package>{<Module1>,<Module2>}`` syntax.
- Enable use of system installation of Sphinx extensions.
- The ``ImportTools`` module is by default no longer included to reduce configure time.
- New ``basis_add_doc`` Doxygen options: ``PREDEFINED``, ``EXCLUDE_BASIS_MODULES``, ``EXCLUDE_BASIS_UTILITIES``.
- Allow generator expressions in list of source files and include directories.
- Updated included GTest and GMock source files.
- Download release packages from GitHub directly during bootstrapping.
- Fixed import of project modules from project configuration file.
- Fixed update of script configuration cache and build script.
- Fixed re-inclusion of package use file when additional components are needed/discovered.
- Fixed missing dependencies in module dependency DAG analysis.
- Removed warning about missing ``CTestConfig.cmake`` file.


.. _v3.2.0:

Release 3.2.0 (Oct 16, 2015)
----------------------------

- Compatibility with CMake 3.
- Compatibility with Python 3.
- Enable build of shared libraries from MATLAB source files.
- Bug fixes.


.. _v3.1.0:

Release 3.1.0 (Mar 28, 2014)
----------------------------

- Custom project layout by overriding the default directory names.
- Revised concept of project modules vs. subprojects.
- Enable build of each module by default.
- Enhanced documentation of project modularization.
- Added project configuraton how-to detailing the various configuration files.
- Allow - and _ in project name.
- Allow use of characters -, _, and + in target names.
- Various improvements to reduce build configuration time.
- Require explicit declaration of link dependency to BASIS Utilities.
- Make RPATH setting optional.
- Root documentation files such as README and COPYING file are now optional.
- Conversion of target name to UID is now disabled by default.
- Added shell script which performs the Quick Start steps.
- Default CMakeList.txt template bootstraps BASIS if no installation found.
- Fixed addition of shared libraries.
- Fixed basis_set_project_property(APPEND PROJECT OtherModule ...).
- Fixed TEST property used by basis_build_script_library.
- Fixed help output of basisproject command.


.. _v3.0.0:

Release 3.0.0 (Jan 15, 2014)
----------------------------

- Slight project name change to "CMake BASIS" (yet usually just referred to as BASIS).
- Generalized templates allow any organization to use BASIS.
- All SBIA specific functionality has been separated out and is now optional.
- Customizable project templates and substitution parameters.
- Default project template configurable during installation using the ``DEFAULT_TEMPLATE`` CMake option.
- Support for binary project template files.
- Multiple new Sphinx themes available for HTML documentation.
- Updated and improved introductory presentation slides.
- Improved and easier to read documentation.
- Hosting new web site of Open Source CMake BASIS project on GitHub.

.. _v2.1.4:

Release 2.1.4 (Apr 20, 2013)
----------------------------


src/tools/which.py
# Copyright (c) 2002-2007 ActiveState Software Inc.
#
# Permission is hereby granted, free of charge, to any person obtaining a
# copy of this software and associated documentation files (the
# "Software"), to deal in the Software without restriction, including
# without limitation the rights to use, copy, modify, merge, publish,
# distribute, sublicense, and/or sell copies of the Software, and to
# permit persons to whom the Software is furnished to do so, subject to
# the following conditions:
#
# The above copyright notice and this permission notice shall be included
# in all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
# OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
# IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
# CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
# SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
#
# Author:
#   Trent Mick (TrentM@ActiveState.com)
# Home:
#   http://trentm.com/projects/which/

r"""Find the full path to commands.

which(command, path=None, verbose=0, exts=None)
    Return the full path to the first match of the given command on the
    path.

whichall(command, path=None, verbose=0, exts=None)
    Return a list of full paths to all matches of the given command on
    the path.

whichgen(command, path=None, verbose=0, exts=None)
    Return a generator which will yield full paths to all matches of the
    given command on the path.
    
By default the PATH environment variable is searched (as well as, on
Windows, the AppPaths key in the registry), but a specific 'path' list
to search may be specified as well.  On Windows, the PATHEXT environment
variable is applied as appropriate.

If "verbose" is true then a tuple of the form
    (<fullpath>, <matched-where-description>)
is returned for each match. The latter element is a textual description
of where the match was found. For example:
    from PATH element 0
    from HKLM\SOFTWARE\...\perl.exe
"""

_cmdlnUsage = """
    Show the full path of commands.

    Usage:
        which [<options>...] [<command-name>...]

    Options:
        -h, --help      Print this help and exit.
        -V, --version   Print the version info and exit.

        -a, --all       Print *all* matching paths.
        -v, --verbose   Print out how matches were located and
                        show near misses on stderr.
        -q, --quiet     Just print out matches. I.e., do not print out
                        near misses.

        -p <altpath>, --path=<altpath>
                        An alternative path (list of directories) may
                        be specified for searching.
        -e <exts>, --exts=<exts>
                        Specify a list of extensions to consider instead
                        of the usual list (';'-separate list, Windows
                        only).

    Show the full path to the program that would be run for each given
    command name, if any. Which, like GNU's which, returns the number of
    failed arguments, or -1 when no <command-name> was given.

    Near misses include duplicates, non-regular files and (on Un*x)
    files without executable access.
"""

__revision__ = "Id: which.py 1448 2007-02-28 19:13:06Z trentm"
__version_info__ = (1, 1, 3)
__version__ = '.'.join(map(str, __version_info__))
__all__ = ["which", "whichall", "whichgen", "WhichError"]

import os
import sys
import getopt
import stat


#---- exceptions

class WhichError(Exception):
    pass



#---- internal support stuff

def _getRegisteredExecutable(exeName):
    """Windows allow application paths to be registered in the registry."""
    registered = None
    if sys.platform.startswith('win'):
        if os.path.splitext(exeName)[1].lower() != '.exe':
            exeName += '.exe'
        import _winreg
        try:
            key = "SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\App Paths\\" +\
                  exeName
            value = _winreg.QueryValue(_winreg.HKEY_LOCAL_MACHINE, key)
            registered = (value, "from HKLM\\"+key)
        except _winreg.error:
            pass
        if registered and not os.path.exists(registered[0]):
            registered = None
    return registered

def _samefile(fname1, fname2):
    if sys.platform.startswith('win'):
        return ( os.path.normpath(os.path.normcase(fname1)) ==\
            os.path.normpath(os.path.normcase(fname2)) )
    else:
        return os.path.samefile(fname1, fname2)

def _cull(potential, matches, verbose=0):
    """Cull inappropriate matches. Possible reasons:
        - a duplicate of a previous match
        - not a disk file
        - not executable (non-Windows)
    If 'potential' is approved it is returned and added to 'matches'.
    Otherwise, None is returned.
    """
    for match in matches:  # don't yield duplicates
        if _samefile(potential[0], match[0]):
            if verbose:
                sys.stderr.write("duplicate: %s (%s)\n" % potential)
            return None
    else:
        if not stat.S_ISREG(os.stat(potential[0]).st_mode):
            if verbose:
                sys.stderr.write("not a regular file: %s (%s)\n" % potential)
        elif sys.platform != "win32" \
             and not os.access(potential[0], os.X_OK):
            if verbose:
                sys.stderr.write("no executable access: %s (%s)\n"\
                                 % potential)
        else:
            matches.append(potential)
            return potential

        
#---- module API

def whichgen(command, path=None, verbose=0, exts=None):
    """Return a generator of full paths to the given command.
    
    "command" is a the name of the executable to search for.
    "path" is an optional alternate path list to search. The default it
        to use the PATH environment variable.
    "verbose", if true, will cause a 2-tuple to be returned for each
        match. The second element is a textual description of where the
        match was found.
    "exts" optionally allows one to specify a list of extensions to use
        instead of the standard list for this system. This can
        effectively be used as an optimization to, for example, avoid
        stat's of "foo.vbs" when searching for "foo" and you know it is
        not a VisualBasic script but ".vbs" is on PATHEXT. This option
        is only supported on Windows.

    This method returns a generator which yields either full paths to
    the given command or, if verbose, tuples of the form (<path to
    command>, <where path found>).
    """
    matches = []
    if path is None:
        usingGivenPath = 0
        path = os.environ.get("PATH", "").split(os.pathsep)
        if sys.platform.startswith("win"):
            path.insert(0, os.curdir)  # implied by Windows shell
    else:
        usingGivenPath = 1

    # Windows has the concept of a list of extensions (PATHEXT env var).
    if sys.platform.startswith("win"):
        if exts is None:
            exts = os.environ.get("PATHEXT", "").split(os.pathsep)
            # If '.exe' is not in exts then obviously this is Win9x and
            # or a bogus PATHEXT, then use a reasonable default.
            for ext in exts:
                if ext.lower() == ".exe":
                    break
            else:
                exts = ['.CMD', '.COM', '.EXE', '.BAT']
        elif not isinstance(exts, list):
            raise TypeError("'exts' argument must be a list or None")
    else:
        if exts is not None:
            raise WhichError("'exts' argument is not supported on "\
                             "platform '%s'" % sys.platform)
        exts = []

    # File name cannot have path separators because PATH lookup does not
    # work that way.
    if os.sep in command or os.altsep and os.altsep in command:
        if os.path.exists(command):
            match = _cull((command, "explicit path given"), matches, verbose)
            if verbose:
                yield match
            else:
                yield match[0]
    else:
        for i in range(len(path)):
            dirName = path[i]
            # On windows the dirName *could* be quoted, drop the quotes
            if sys.platform.startswith("win") and len(dirName) >= 2\
               and dirName[0] == '"' and dirName[-1] == '"':
                dirName = dirName[1:-1]
            for ext in ['']+exts:
                absName = os.path.abspath(
                    os.path.normpath(os.path.join(dirName, command+ext)))
                if os.path.isfile(absName):
                    if usingGivenPath:
                        fromWhere = "from given path element %d" % i
                    elif not sys.platform.startswith("win"):
                        fromWhere = "from PATH element %d" % i
                    elif i == 0:
                        fromWhere = "from current directory"
                    else:
                        fromWhere = "from PATH element %d" % (i-1)
                    match = _cull((absName, fromWhere), matches, verbose)
                    if match:
                        if verbose:
                            yield match
                        else:
                            yield match[0]
        match = _getRegisteredExecutable(command)
        if match is not None:
            match = _cull(match, matches, verbose)
            if match:
                if verbose:
                    yield match
                else:
                    yield match[0]


def which(command, path=None, verbose=0, exts=None):
    """Return the full path to the first match of the given command on
    the path.
    
    "command" is a the name of the executable to search for.
    "path" is an optional alternate path list to search. The default it
        to use the PATH environment variable.
    "verbose", if true, will cause a 2-tuple to be returned. The second
        element is a textual description of where the match was found.
    "exts" optionally allows one to specify a list of extensions to use
        instead of the standard list for this system. This can
        effectively be used as an optimization to, for example, avoid
        stat's of "foo.vbs" when searching for "foo" and you know it is
        not a VisualBasic script but ".vbs" is on PATHEXT. This option
        is only supported on Windows.

    If no match is found for the command, a WhichError is raised.
    """
    try:
        match = whichgen(command, path, verbose, exts).next()
    except StopIteration:
        raise WhichError("Could not find '%s' on the path." % command)
    return match


def whichall(command, path=None, verbose=0, exts=None):
    """Return a list of full paths to all matches of the given command
    on the path.  

    "command" is a the name of the executable to search for.
    "path" is an optional alternate path list to search. The default it
        to use the PATH environment variable.
    "verbose", if true, will cause a 2-tuple to be returned for each
        match. The second element is a textual description of where the
        match was found.
    "exts" optionally allows one to specify a list of extensions to use
        instead of the standard list for this system. This can
        effectively be used as an optimization to, for example, avoid
        stat's of "foo.vbs" when searching for "foo" and you know it is
        not a VisualBasic script but ".vbs" is on PATHEXT. This option
        is only supported on Windows.
    """
    return list( whichgen(command, path, verbose, exts) )



#---- mainline

def main(argv):
    all = 0
    verbose = 0
    altpath = None
    exts = None
    try:
        optlist, args = getopt.getopt(argv[1:], 'haVvqp:e:',
            ['help', 'all', 'version', 'verbose', 'quiet', 'path=', 'exts='])
    except getopt.GetoptError, msg:
        sys.stderr.write("which: error: %s. Your invocation was: %s\n"\
                         % (msg, argv))
        sys.stderr.write("Try 'which --help'.\n")
        return 1
    for opt, optarg in optlist:
        if opt in ('-h', '--help'):
            print _cmdlnUsage
            return 0
        elif opt in ('-V', '--version'):
            print "which %s" % __version__
            return 0
        elif opt in ('-a', '--all'):
            all = 1
        elif opt in ('-v', '--verbose'):
            verbose = 1
        elif opt in ('-q', '--quiet'):
            verbose = 0
        elif opt in ('-p', '--path'):
            if optarg:
                altpath = optarg.split(os.pathsep)
            else:
                altpath = []
        elif opt in ('-e', '--exts'):
            if optarg:
                exts = optarg.split(os.pathsep)
            else:
                exts = []

    if len(args) == 0:
        return -1

    failures = 0
    for arg in args:
        #print "debug: search for %r" % arg
        nmatches = 0
        for match in whichgen(arg, path=altpath, verbose=verbose, exts=exts):
            if verbose:
                print "%s (%s)" % match
            else:
                print match
            nmatches += 1
            if not all:
                break
        if not nmatches:
            failures += 1
    return failures


if __name__ == "__main__":
    sys.exit( main(sys.argv) )


